find/client/http/provider_cache.go

```diff
@@ -0,0 +1,194 @@
+	}
+
+	var providerInfo model.ProviderInfo
+	err = json.Unmarshal(body, &providerInfo)
+	if err != nil {
+		return nil, err
+	}
+	return &providerInfo, nil
+}
+
+func (pc *providerCache) getResults(ctx context.Context, pid peer.ID, ctxID []byte, metadata []byte) ([]model.ProviderResult, error) {
+	wrapper := pc.getPInfoWrapper(pid)
+
+	// If ProviderInfo isn't in the cache or if the record has expired - try to
+	// fetch a new ProviderInfo and update the cache
+	if wrapper == nil || time.Since(wrapper.ts) > pc.ttl {
+		pinfo, err := pc.fetchPInfo(ctx, pid)
+		if err != nil {
+			return nil, err
+		}
+
+		wrapper = &pinfoWrapper{
+			ts:    time.Now(),
+			pinfo: pinfo,
+			cxps:  make(map[string]*model.ContextualExtendedProviders),
+		}
+		if pinfo.ExtendedProviders != nil {
+			for _, cxp := range pinfo.ExtendedProviders.Contextual {
+				wrapper.cxps[cxp.ContextID] = &cxp
+			}
+		}
+		pc.setPInforWrapper(pid, wrapper)
+	}
+
+	results := make([]model.ProviderResult, 0, 1)
+
+	results = append(results, model.ProviderResult{
+		ContextID: ctxID,
+		Metadata:  metadata,
+		Provider:  &wrapper.pinfo.AddrInfo,
+	})
+
+	// return results if there are no further extended providers to unpack
+	if wrapper.pinfo.ExtendedProviders == nil {
+		return results, nil
+	}
+
+	// If override is set to true at the context level then the chain
+	// level EPs should be ignored for this context ID
+	override := false
+
+	// Adding context-level EPs if they exist
+	if contextualEpRecord, ok := wrapper.cxps[string(ctxID)]; ok {
+		override = contextualEpRecord.Override
+		for i, xpinfo := range contextualEpRecord.Providers {
+			xmd := contextualEpRecord.Metadatas[i]
+			// Skippng the main provider's record if its metadata is nil or is
+			// the same as the one retrieved from the indexer, because such EP
+			// record does not advertise any new protocol.
+			if xpinfo.ID == wrapper.pinfo.AddrInfo.ID &&
+				(len(xmd) == 0 || bytes.Equal(xmd, metadata)) {
+				continue
+			}
+			// Use metadata from advertisement if one hasn't been specified for
+			// the extended provider
+			if xmd == nil {
+				xmd = metadata
+			}
+
+			results = append(results, model.ProviderResult{
+				ContextID: ctxID,
+				Metadata:  xmd,
+				Provider:  &xpinfo,
+			})
+		}
+	}
+
+	// If override is true then don't include chain-level EPs
+	if override {
+		return results, nil
+	}
+
+	// Adding chain-level EPs if such exist
+	for i, xpinfo := range wrapper.pinfo.ExtendedProviders.Providers {
+		xmd := wrapper.pinfo.ExtendedProviders.Metadatas[i]
+		// Skippng the main provider's record if its metadata is nil or is the
+		// same as the one retrieved from the indexer, because such EP record
+		// does not advertise any new protocol.
+		if xpinfo.ID == wrapper.pinfo.AddrInfo.ID &&
+			(len(xmd) == 0 || bytes.Equal(xmd, metadata)) {
+			continue
+		}
+		// Use metadata from advertisement if one hasn't been specified for the
+		// extended provider
+		if xmd == nil {
```

Comment:
Would it be safer to say
```suggestion
		if len(xmd) == 0 {
```
In case there is an empty byte slice?